lib/aliquot/payment.rb

```diff
@@ -0,0 +1,135 @@
+module Aliquot
+  ##
+  # A Payment represents a single payment using Google Pay.
+  # It is used to verify/decrypt the supplied token by using the shared secret,
+  # thus avoiding having knowledge of merchant primary keys.
+  class Payment
+    ##
+    # Parameters:
+    # token_string::  Google Pay token (JSON string)
+    # shared_secret:: Base64 encoded shared secret
+    # merchant_id::   Google Pay merchant ID ("merchant:<SOMETHING>")
+    # logger::        The logger to use. Default: Logger.new($stdout)
+    # signing_keys::  Formatted list of signing keys used to sign token contents.
+    #                 Otherwise a thread continuously updating google signing
+    #                 keys will be started.
+    def initialize(token_string, shared_secret, merchant_id, logger: Logger.new($stdout), signing_keys: nil)
+      Aliquot.start_key_updater(logger) if $key_updater_thread.nil? && signing_keys.nil?
+
+      validation = Aliquot::Validator::Token.new(JSON.parse(token_string))
+      validation.validate
+
+      @token = validation.output
+
+      @shared_secret = shared_secret
+      @merchant_id   = merchant_id
+      @signing_keys  = signing_keys
+    end
+
+    ##
+    # Validate and decrypt the token.
+    def process
+      unless valid_protocol_version?
+        raise Error, 'only ECv1 protocolVersion is supported'
+      end
+
+      unless valid_signature?
+        raise InvalidSignatureError
+      end
+
+      validator = Aliquot::Validator::SignedMessage.new(JSON.parse(@token['signedMessage']))
+      validator.validate
+      signed_message = validator.output
+
+      aes_key, mac_key = derive_keys(signed_message['ephemeralPublicKey'], @shared_secret, 'Google')
+
+      unless valid_mac?(mac_key, signed_message['encryptedMessage'], signed_message['tag'])
+        raise InvalidMacError
+      end
+
+      @message = JSON.parse(decrypt(aes_key, signed_message['encryptedMessage']))
+
+      Aliquot::Validator::EncryptedMessageValidator.new(@message).validate
```

Comment:
We could consider having a "tinier" validator to just ensure that we can extract the `messageExpiration` part to decide if the message is expired and then leave it to the user to validate the remaining of the message. WDYT? @kse-clearhaus?

Reply:
I'm not sure I understand.

You suggest ignoring everything and just decrypting and checking `messageExpiration`?

Reply:
My thinking is if the user of aliquot should themselves handle the validation of the resulting message (as it is the case with pedicel) or if aliquot should enforce the validation.

Reply:
Ah, yes.

It is absolutely prettier to check payment credentials  e.g. inside Klovedal.
The validator, in essence, just needs to check existance and expected type.

Some data massaging might be necessary, e.g. expiry information from ints to strings.